src/database.py

```diff
@@ -0,0 +1,144 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""The Jenkins agent relation observer."""
+import logging
+import typing
+
+import psycopg2
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+from ops.charm import CharmBase
+from ops.framework import Object
+from psycopg2 import sql
+from psycopg2.extensions import connection
+
+logger = logging.getLogger(__name__)
+
+
+class DatabaseObserver(Object):
+    """The Database relation observer."""
+
+    _RELATION_NAME = "database"
+
+    def __init__(self, charm: CharmBase):
+        """Initialize the observer and register event handlers.
+
+        Args:
+            charm: The parent charm to attach the observer to.
+        """
+        super().__init__(charm, "database-observer")
+        self._charm = charm
+        # SUPERUSER is required to update pg_database
+        self.database = DatabaseRequires(
+            self._charm,
+            relation_name=self._RELATION_NAME,
+            database_name=self._charm.app.name,
+            extra_user_roles="SUPERUSER",
+        )
+
+    def get_relation_data(self) -> typing.Optional[typing.Dict]:
```

Comment:
Similar to earlier comment, should this be done via the charm state?

Reply:
Yes, changed.